### ISSUE TITLE
Regtesting Improvements

### ISSUE THREAD

**Issue by oschuett** [cp2k/cp2k#2996]
I'd like to make a number of improvements to our regtesting script:

1. Give [test types](https://github.com/cp2k/cp2k/blob/master/tests/TEST_TYPES) meaningful identifiers instead of just  numbers.

2. Allow to check also other outputs besides the stdout.

3. Allow for multiple checks after running an input file.

As a first step, I'd like to move the [do_regtest.py](https://github.com/cp2k/cp2k/blob/master/tools/regtesting/do_regtest.py) script from `cp2k/tools/regtesting` to `cp2k/tests`.

Are there any comments or concerns?

**Comment by fstein93** [comment]
I am in favor of your ideas as it allows us to reduce the number of calls and thus runtime of the test suite to CP2K whereas you second idea allows us to test common mistakes in input files of CP2K assuming that we can then check them improving the coverage.

**Comment by juerghutter** [comment]
What about to add a new macro ?

CPTEST(TAG, VALUE)


________________________________________

**Comment by oschuett** [comment]
> What about to add a new macro ? `CPTEST(TAG, VALUE)`

What would be the advantage over our current pattern matching?

These patterns are surprisingly stable. The last time one had to be update was [more than two years ago](https://github.com/cp2k/cp2k/commit/0c4637b047664bdb0e05708583f3c20750a15988#diff-946efca8ef1f8b3ef9e820ad626c632a0812bb34ba8babca9825b71e21f2dc7e).

**Comment by juerghutter** [comment]
This would allow for a single pattern rule (with a changing tag) instead
of 100 different rules.
It is transparent for the developer and easy to handle.
It could run with the current scheme in parallel (adds 1 new rule).

________________________________________

**Comment by oschuett** [comment]
Ok, now I see where you're going. You want to eliminate the [TEST_TYPES](https://github.com/cp2k/cp2k/blob/master/tests/TEST_TYPES) entirely. That is indeed tempting.

However, I see two downsides:

1. The tagged output will clutter the regular output. We could collected it and print everything at the very end, but then we'd have nothing in case of a midway crash.
2. There is no way to migrated the existing tests automatically. Hence, the TEST_TYPES file will stay around for the foreseeable future and we'll have to deal with [two systems](https://xkcd.com/927/).

I agree that adding a new test pattern is  a bit  cumbersome. I'm hoping to alleviate this by moving to a more familiar syntax, i.e. changing `TEST_TYPES` to `test_types.py`.

**Comment by mkrack** [comment]
I think Jürg's proposal is worth a try, since it has some pros, If I got his idea right:

- It would allow performing several tests (tags) concurrently with a single test job which would reduce the number of test runs.
- Changes in the output format would not interfere with the regtesting any longer.
- It could also be used for collecting key results from a run (like the dump section at the end of a Gaussian output)
- The new scheme could be combined with the current scheme.

> 1. The tagged output will clutter the regular output. We could collected it and print everything at the very end, but then we'd have nothing in case of a midway crash.

The tagged output could be optionally activated by a compile flag like `-D__REGTEST`

> 2. There is no way to migrated the existing tests automatically. Hence, the TEST_TYPES file will stay around for the foreseeable future and we'll have to deal with [two systems](https://xkcd.com/927/).

Thats's true and a price to pay, but if you are revising `TEST_TYPES` anyway, a new more flexible scheme could already be considered for its design.

**Comment by oschuett** [comment]
> - It would allow performing several tests (tags) concurrently with a single test job which would reduce the number of test runs.

Yes, I'll definitely add support for that by allowing consecutive lines in a `TEST_FILES` to use the same input file.

> - Changes in the output format would not interfere with the regtesting any longer.

Since our output format hardly changes in reality, I'd deem this only a small advantage.

> - It could also be used for collecting key results from a run (like the dump section at the end of a Gaussian output)

I agree, this would be a useful feature on its own. It's essentially #215.

> The tagged output could be optionally activated by a compile flag like -D__REGTEST

I don't know. I'd rather not give the code an opportunity to behave differently when being tested.

> ...if you are revising TEST_TYPES anyway, a new more flexible scheme could already be considered for its design.

The new `test_types.py` will be very extensible. So if someone builds #215 then we can readily tie it into our testing.

**Comment by mkrack** [comment]
If testing and the collection of key results will become a combined feature, then there is, of course, no need for a `-D__REGTEST` compile flag.

**Comment by oschuett** [comment]
In case anybody was wondering what happened to this project: I got stuck on assigning meaningful identifiers to our [test types](https://github.com/cp2k/cp2k/blob/master/tests/TEST_TYPES).

It turns out that determining the meaning of each of our 110+ types takes quite a bit of time. So any help would be very welcome.

Essentially, I need a list like this:

| old test type number | new test type name | short description                   |
| -------------------- | ------------------ | ----------------------------------- |
| 1                    | E_total            | Total energy reported by force eval |
| 2                    | E_pot_md           | Potential energy of MD              |
| 3                    | ...                | ...                                 |

**Comment by mtaillefumier** [comment]
Another thing. Would it be possible to actually not have `$(mpiexec) -n` hardcoded but something like `$(launch_command) $(cp2k_exe)` with no other parameters than the input and ouput files. It would help a lot running regression tests on clusters that do not use slurm for instance.

**Comment by mtaillefumier** [comment]
> In case anybody was wondering what happened to this project: I got stuck on assigning meaningful identifiers to our [test types](https://github.com/cp2k/cp2k/blob/master/tests/TEST_TYPES).
> 
> It turns out that determining the meaning of each of our 110+ types takes quite a bit of time. So any help would be very welcome.
> 
> Essentially, I need a list like this:
> old test type number 	new test type name 	short description
> 1 	E_total 	Total energy reported by force eval
> 2 	E_pot_md 	Potential energy of MD
> 3 	... 	...

it happens that these projects get delayed for some reasons. Do not feel bad about this.

**Comment by oschuett** [comment]
> Another thing. Would it be possible to actually not have `$(mpiexec) -n` hardcoded but something like `$(launch_command) $(cp2k_exe)` with no other parameters than the input and ouput files. It would help a lot running regression tests on clusters that do not use slurm for instance.

That shouldn't be too hard to add, however I'd like to understand you use case a little better first.
Since `mpiexec -n` is part of the MPI standard, I'm surprised it doesn't always work. Maybe you could elaborate a bit?

> it happens that these projects get delayed for some reasons. Do not feel bad about this.

Thanks 😊

**Comment by alazzaro** [comment]
> > Another thing. Would it be possible to actually not have `$(mpiexec) -n` hardcoded but something like `$(launch_command) $(cp2k_exe)` with no other parameters than the input and ouput files. It would help a lot running regression tests on clusters that do not use slurm for instance.
> 
> That shouldn't be too hard to add, however I'd like to understand you use case a little better first. Since `mpiexec -n` is part of the MPI standard, I'm surprised it doesn't always work. Maybe you could elaborate a bit?

For systems with batch systems (example SLURM), you don't have mpiexec, basically the batch system starts MPI with a PMI (example, srun with slurm).

**Comment by oschuett** [comment]
But `srun` also supports the `-n` option: https://slurm.schedmd.com/srun.html#OPT_ntasks

While one can leave it up to `srun` to pick the number of MPI ranks dynamically, I'd strongly advice against it because the `do_regtest.py` script uses that number for several other decision:

- setting the number of [parallel workers](https://github.com/cp2k/cp2k/blob/55a83f1f3aa5ca1b07fcaa8d4512f2f0554257e5/tests/do_regtest.py#L254)

- checking each [test's requirements](https://github.com/cp2k/cp2k/blob/55a83f1f3aa5ca1b07fcaa8d4512f2f0554257e5/tests/do_regtest.py#L156)

- setting the [visible GPU devices](https://github.com/cp2k/cp2k/blob/55a83f1f3aa5ca1b07fcaa8d4512f2f0554257e5/tests/do_regtest.py#L304)

**Comment by alazzaro** [comment]
I have to admit, I used it already, and it works. My comment was to try to clarify @mtaillefumier request

**Comment by mtaillefumier** [comment]
I never said the option should disappear. What I suggested is to not assume that -n means the number of tasks by default. 

Something like this would work 

default value :

`mpi_command = "mpiexec -n {numtask_task} {cp2k_exe}"`

```python
mpi_command_exe = Template(mpi_command)
submit(mpi_command_exe.substitute(mpi_command=mpi_command, num_tasks=num_tasks, cp2k_exe=my_cp2k))

```

the script would have a default but also would allow a custom configuration. I had to modify the script to make it "work" on summit before I litteraly went back in time to retrieve the shell script.
